pymc/sampling/forward.py

```diff
@@ -497,22 +499,148 @@ def sample_posterior_predictive(
 
     Examples
     --------
-    Thin a sampled inferencedata by keeping 1 out of every 5 draws
-    before passing it to sample_posterior_predictive
+    The behavior of ``var_names``
+    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    The function returns posterior predictive samples of any variable included in ``var_names``,
+    conditioned on the values of other random variables found in the trace.
+    To ensure the samples are internally consistent, any random variable that depends
+    on another random variable that is being sampled is itself sampled.
+    The final list of variables being sampled is shown in the log output.
+
+    Note that if a random variable has no dependency on other random variables,
+    posterior predictive sampling is equivalent to prior sampling.
+    Likewise, if all random variables are being sampled, the behavior is equivalent to
+    that of :func:`~pymc.sample_prior_predictive`.
```

Comment:
Can we create a `UserWarning` to warn users when a a variable was sampled from the prior? I feel like this would help tremendously in disambiguating what `sample_posterior_predictive` is doing. Because right now it's doing it implicitly

Reply:
I don't think users ever want (as in 99.9% of the time) an unobserved variable to be resampled regardless of whether that's equivalent to the prior or slightly different because it depends on some non volatile variable.

I would rather go with my suggestion in https://github.com/pymc-devs/pymc/issues/7069#issuecomment-1878388607 which fails explicitly unless you manually add the variable to `sample_vars`

Reply:
> I would rather go with my suggestion in https://github.com/pymc-devs/pymc/issues/7069#issuecomment-1878388607 which fails explicitly unless you manually add the variable to sample_vars

That's fine by me too. My point is that this behavior should be explicit, to avoid surprising hidden behavior, so making it fail unless the user asks for it sounds great to me!

Bonus question though: what do you mean by "(non) volatile variable"? I've seen it mentioned several time on this PR, but I haven't seen it defined.

Reply:
> That's fine by me too. My point is that this behavior should be explicit, to avoid surprising hidden behavior, so making it fail unless the user asks for it sounds great to me!

Sure but I'm not tackling this in this PR because it will probably be a breaking change and require consensus. This PR only documents the existing behavior and as such can't be really rejected (besides whether the examples are correct/clear).

Reply:
> Bonus question though: what do you mean by "(non) volatile variable"? I've seen it mentioned several time on this PR, but I haven't seen it defined.

I forgot to define it. Volatile variables are variables that depend on variables being resampled (or coords/mutable data that has changed) and as such must also be resampled even if they existed in the trace and where not included in `var_names`, to ensure all draws are jointly consistent

Reply:
> Sure but I'm not tackling this in this PR because it will probably be a breaking change and require consensus. This PR only documents the existing behavior and as such can't be really rejected (besides whether the examples are correct/clear).
> 
> I also think this will help with the discussion because right now most people don't even know how it works

:100 on this. I think we should merge this PR ASAP, to 1) make things clearer ASAP for users, 2) open that discussion for the breaking change and move towards solving it.
Do you need any help on this? Beside reviewing consistently.

Thanks for the definition of volatile variables. Much clearer now 👌